{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Differentially expressed genes as identified in cells with cryptic| mitochondrial mutations\n",
    "\n",
    "We use single-cell RNA-sequencing data from Enge et al. to demonstrate that we can identify DEGs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import os\n",
    "import argparse\n",
    "import pickle as pkl\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy.stats as stats\n",
    "\n",
    "import seaborn as sns\n",
    "\n",
    "sns.set_context(\"paper\")\n",
    "sns.set_style(\"white\")\n",
    "sns.set(font_scale=1.2) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 0. Data loading and preparation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the dataframe with mutation information\n",
    "varianceData = pkl.load( open( \"./../../data/precomputedData/Enge/engeVariantsSTAR200.pkl\", \"rb\" ) )\n",
    "# expression matrix\n",
    "adata = sc.read_h5ad(\"./../../data/precomputedData/Enge/engeFilteredExpression200.h5ad\")\n",
    "# metadata\n",
    "metadata = pkl.load( open( \"./../../data/precomputedData/Enge/enge_metadata.pkl\", \"rb\" ) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add age and cell type as annotations to the expression matrix \n",
    "# get all ages from the variance data\n",
    "age = [ metadata[metadata['SRR_id'] == item]['donor_age'].iloc[0] if (item in metadata['SRR_id'].values) else 0 for item in adata.obs_names]\n",
    "# add it to the expression matrix\n",
    "adata.obs['donor_age'] = age\n",
    "# cellType = [ metadata[metadata['SRR_id'] == item]['inferred_cell_type'].iloc[0] if (item in metadata['SRR_id'].values) else 0 for item in adata.obs_names]\n",
    "# adata.obs['cell_type'] = cellType"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# normalisation\n",
    "sc.pp.normalize_total(adata,target_sum=1e6)\n",
    "sc.pp.log1p(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define a function that computes mitochondrial load and adds it to the data frame\n",
    "def computeMitochondrialLoad(expressionMatrix,varianceDataFrame,\n",
    "                             threshold = 0.05, #threshold: heteroplasmy threshold (only mutations above this threshold are considered)\n",
    "                             mutantType = 'all', # which type of mutations should be considered, list of strings, e.g., ['Common mutation','Developmental']\n",
    "                             removeSynonymous = 0, # whether to remove synonymous mutations from the mutations\n",
    "                             pathologyType = 'all', # which type of pathology we look at\n",
    "                             subsetCells = 'all', # whether to keep only certain type of cells\n",
    "                             rescaling= False, # rescaling by considering the number of bases passed (not recommended)\n",
    "                             muName = 'MU' # name given to the mitochondrial load in the expression matrix Anndata object\n",
    "                            ): \n",
    "    \n",
    "    # Apply heteroplasmy threshold \n",
    "    varianceDataFrame = varianceDataFrame[threshold<varianceDataFrame['HF']]\n",
    "    \n",
    "    # Keep only certain type of mutations\n",
    "    if (mutantType == 'all'):\n",
    "        pass\n",
    "    else:\n",
    "        # keep only variance which is in this list\n",
    "         varianceDataFrame = varianceDataFrame[varianceDataFrame['mutant_type'].isin(mutantType)]\n",
    "    \n",
    "    # Keep only synonomous or non-synonomous mutation\n",
    "    if removeSynonymous == True:\n",
    "        varianceDataFrame= varianceDataFrame[~(varianceDataFrame['MutPred_Prediction'] == 'Synonymous')]\n",
    "    elif (removeSynonymous == -1):\n",
    "        # keep only synonomous mutations\n",
    "        varianceDataFrame= varianceDataFrame[(varianceDataFrame['MutPred_Prediction'] == 'Synonymous')]\n",
    "    \n",
    "    # Keep only mutations of a certain pathology\n",
    "    if (pathologyType == 'all'):\n",
    "        pass\n",
    "    else:\n",
    "        # keep only variance which is in this list\n",
    "         varianceDataFrame = varianceDataFrame[varianceDataFrame['MutPred_Prediction'].isin(pathologyType)]\n",
    "    \n",
    "    # Keep only certain cells\n",
    "    if (subsetCells == 'all'):\n",
    "        pass\n",
    "    else:\n",
    "        # keep only variance which is in this list\n",
    "         varianceDataFrame = varianceDataFrame[varianceDataFrame['sample_id'].isin(subsetCells)]\n",
    "    \n",
    "    # optional rescaling (it is easier to do this before the summation and the math works out)\n",
    "    if (rescaling == True):\n",
    "        varianceDataFrame['scalingFactor'] = 16569/varianceDataFrame['bases_passed']\n",
    "        #varianceDataFrame['scalingFactor'][np.isinf(varianceDataFrame['scalingFactor'])] = 0\n",
    "        varianceDataFrame['HF'] = varianceDataFrame.HF * varianceDataFrame.scalingFactor\n",
    "        \n",
    "\n",
    "    # compute the mitochondrial load\n",
    "    groupedData = varianceDataFrame.groupby('sample_id') # group for each cell\n",
    "    mu = groupedData.sum()['HF'] # mitochondrial load is then the sum over the heteroplasmies\n",
    "    \n",
    "    \n",
    "    # add the mitoLoad as observable MU\n",
    "    expressionMatrix.obs[muName] = mu\n",
    "    expressionMatrix.obs[muName] = expressionMatrix.obs[muName] .fillna(0)\n",
    "    \n",
    "    return(expressionMatrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# DEG calculation \n",
    "\n",
    "def findDEGMitoLoad(expressionMatrix,varianceDataFrame,\n",
    "                    threshold = 0.05, #threshold: heteroplasmy threshold (only mutations above this threshold are considered)\n",
    "                    mutantType = 'all', # which type of mutations should be considered, list of strings, e.g., ['Common mutation','Developmental']\n",
    "                    removeSynonymous = 0, # whether to remove synonymous mutations from the mutations\n",
    "                    pathologyType = 'all', # which type of pathology we look at\n",
    "                    subsetCells = 'all', # whether to keep only certain type of cells\n",
    "                    muThreshold = 0.0, # threshold to divide cells into with or without mutation\n",
    "                    significanceLevel=0.05, # if zero, all genes are returned\n",
    "                    rescaling= False): # rescaling by considering the number of bases passed (not recommended)):\n",
    "    \n",
    "    # compute the mitochondrial load\n",
    "    expressionMatrixMu = computeMitochondrialLoad(expressionMatrix,varianceDataFrame,threshold = threshold,mutantType = mutantType, # which type of mutations should be considered, list of strings, e.g., ['Common mutation','Developmental']\n",
    "                    removeSynonymous = removeSynonymous, # whether to remove synonymous mutations from the mutations\n",
    "                    pathologyType = pathologyType, # which type of pathology we look at\n",
    "                    subsetCells = subsetCells, # whether to keep only certain type of cells\n",
    "                    rescaling= rescaling)\n",
    "    \n",
    "    \n",
    "    # we need at least two cells with a mutation and two cells without\n",
    "    nCellsWithMutation = np.sum(expressionMatrixMu.obs['MU'] >muThreshold)\n",
    "    nCells = len(expressionMatrixMu.obs_names)\n",
    "    if (   (nCellsWithMutation>1) & (nCellsWithMutation<nCells-1)  ):\n",
    "        # create a new observable that is binary whether the mitoload is above a threshold\n",
    "        maxMitoLoad = np.max(expressionMatrixMu.obs['MU'])\n",
    "        expressionMatrixMu.obs['mitLoadBinary'] = pd.Categorical(np.ceil(expressionMatrixMu.obs['MU']/maxMitoLoad))\n",
    "        # do the DEG with a wilcoxon\n",
    "        sc.tl.rank_genes_groups(expressionMatrixMu, 'mitLoadBinary', method='wilcoxon',n_genes=expressionMatrixMu.n_vars)\n",
    "        # get out which genes are differentially expressed\n",
    "        DEGresult = expressionMatrixMu.uns['rank_genes_groups']\n",
    "        \n",
    "        if significanceLevel>0:\n",
    "            significantIndexes = np.where(DEGresult['pvals_adj']['1.0']<significanceLevel)\n",
    "\n",
    "            listSignificantGenes = pd.DataFrame()\n",
    "            listSignificantGenes['names'] = DEGresult['names']['1.0'][significantIndexes]\n",
    "            listSignificantGenes['pvals_adj'] = DEGresult['pvals_adj']['1.0'][significantIndexes]\n",
    "            listSignificantGenes['names'] = DEGresult['names']['1.0'][significantIndexes]\n",
    "            listSignificantGenes['logfoldchanges'] = DEGresult['logfoldchanges']['1.0'][significantIndexes]\n",
    "        else: # otherwise return it for all genes\n",
    "            listSignificantGenes = pd.DataFrame()\n",
    "            listSignificantGenes['names'] = DEGresult['names']['1.0']\n",
    "            listSignificantGenes['pvals_adj'] = DEGresult['pvals_adj']['1.0']\n",
    "            listSignificantGenes['pvals'] = DEGresult['pvals']['1.0']\n",
    "            listSignificantGenes['logfoldchanges'] = DEGresult['logfoldchanges']['1.0']\n",
    "        # sort by adjusted p-value\n",
    "        listSignificantGenes = listSignificantGenes.sort_values(by='pvals_adj')\n",
    "    else:\n",
    "        # empty dataframe if not enough mitoload to call\n",
    "        listSignificantGenes = pd.DataFrame()\n",
    "    return(listSignificantGenes)\n",
    "\n",
    "\n",
    "# def DEG_threshold(expressionMatrix,varianceDataFrame,\n",
    "#                   thresholdVec = np.arange(0,1,0.05),\n",
    "#                   mutantTypeVec = ['all'], # which type of mutations should be considered, list of strings, e.g., ['Common mutation','Developmental']\n",
    "#                   removeSynonymousVec = [0], # whether to remove synonymous mutations from the mutations\n",
    "#                   pathologyTypeVec = ['all'], # which type of pathology we look at\n",
    "#                   subsetCells = 'all', # whether to keep only certain type of cells\n",
    "#                   rescalingVec= [False]): # rescaling by considering the number of bases passed (not recommended)):\n",
    "\n",
    "#     outputDf = pd.DataFrame()\n",
    "#     colNames = ['nSig','nSig (upregulated)','nSig (downregulated)','pVal combined','threshold','mutant type','removed synonomous','pathology','rescaling']\n",
    "    \n",
    "#     # go over the heteroplasmy thresholds\n",
    "#     print('thresholds:')\n",
    "#     for t in thresholdVec:\n",
    "#         print(t)\n",
    "#         for m in mutantTypeVec:\n",
    "#             for rSyn in removeSynonymousVec:\n",
    "#                 for pat in pathologyTypeVec:\n",
    "#                     for rescale in rescalingVec:\n",
    "\n",
    "#                         listSignificantGenes = findDEGMitoLoad(expressionMatrix,varianceDataFrame,\n",
    "#                                     threshold = t, #threshold: heteroplasmy threshold (only mutations above this threshold are considered)\n",
    "#                                     mutantType = m, # which type of mutations should be considered, list of strings, e.g., ['Common mutation','Developmental']\n",
    "#                                     removeSynonymous = rSyn, # whether to remove synonymous mutations from the mutations\n",
    "#                                     pathologyType = pat, # which type of pathology we look at\n",
    "#                                     rescaling= rescale)\n",
    "#                         # compute the number of significante genes\n",
    "#                         nSig = listSignificantGenes.shape[0]\n",
    "#                         if nSig>0:\n",
    "#                             nSigUp = np.sum(listSignificantGenes['logfoldchanges']>0) # number of upregulated genes\n",
    "#                             nSigDown = np.sum(listSignificantGenes['logfoldchanges']<0) # number of downregulated genes\n",
    "#                             pValCombined = stats.combine_pvalues(listSignificantGenes['pvals_adj'])[1] # Fisher's method\n",
    "                            \n",
    "#                             df = pd.DataFrame([[nSig,nSigUp,nSigDown,pValCombined,t,m[0],rSyn,pat[0],rescale]], columns=colNames)\n",
    "#                             outputDf = outputDf.append(df)\n",
    "#                         else:\n",
    "#                             # add that no genes are significant\n",
    "#                             df = pd.DataFrame([[0,0,0,1,t,m[0],rSyn,pat[0],rescale]], columns=colNames)\n",
    "#                             outputDf = outputDf.append(df)\n",
    "                        \n",
    "#     return(outputDf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We only look at mutations that could have been detected in at least ten cells\n",
    "varianceSubset = varianceData[varianceData['cells_possible']>10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Compute DEGs for at a different thresholds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run analysis\n",
    "DEGs_Cryptic_10 = findDEGMitoLoad(adata,varianceSubset,threshold=0.10,mutantType=['Cryptic'],rescaling=True,removeSynonymous=1,significanceLevel=0.05)\n",
    "DEGs_Cryptic_30 = findDEGMitoLoad(adata,varianceSubset,threshold=0.3,mutantType=['Cryptic'],rescaling=True,removeSynonymous=1,significanceLevel=0.05)\n",
    "DEGs_Cryptic_40 = findDEGMitoLoad(adata,varianceSubset,threshold=0.40,mutantType=['Cryptic'],rescaling=True,removeSynonymous=1,significanceLevel=0.05)\n",
    "DEGs_Cryptic_95 = findDEGMitoLoad(adata,varianceSubset,threshold=0.95,mutantType=['Cryptic'],rescaling=True,removeSynonymous=1,significanceLevel=0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save\n",
    "DEGs_Cryptic_10.to_csv('./DEG_out/deg_Enge_t10.csv')\n",
    "DEGs_Cryptic_30.to_csv('./DEG_out/deg_Enge_t30.csv')\n",
    "DEGs_Cryptic_40.to_csv('./DEG_out/deg_Enge_t40.csv')\n",
    "DEGs_Cryptic_95.to_csv('./DEG_out/deg_Enge_t95.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
